File: LLD/Design-Patterns/Facade/code/facade.js
/**
 * Facade Pattern — OrderFacade coordinating several subsystems behind one call.
 * Run: node facade.js
 */

class InventoryService {
  reserveItem(item) {
    console.log(`Inventory: reserved '${item}'`);
    return true;
  }
}

class PaymentService {
  charge(amount) {
    console.log(`Payment: charged $${amount}`);
    return true;
  }
}

class ShippingService {
  scheduleShipment(item) {
    console.log(`Shipping: scheduled shipment for '${item}'`);
  }
}

class NotificationService {
  sendConfirmation(item) {
    console.log(`Notification: confirmation email sent for '${item}'`);
  }
}

class OrderFacade {
  constructor() {
    this.inventory = new InventoryService();
    this.payment = new PaymentService();
    this.shipping = new ShippingService();
    this.notification = new NotificationService();
  }

  placeOrder(item, price) {
    console.log(`--- Placing order for ${item} ---`);
    if (this.inventory.reserveItem(item) && this.payment.charge(price)) {
      this.shipping.scheduleShipment(item);
      this.notification.sendConfirmation(item);
      console.log("Order complete.");
    }
  }
}

const orderFacade = new OrderFacade();
orderFacade.placeOrder("Wireless Mouse", 24.99);

module.exports = { OrderFacade };

File: LLD/Design-Patterns/Facade/code/facade.py
"""
Facade Pattern — OrderFacade coordinating several subsystems behind one call.
Run: python facade.py
"""


class InventoryService:
    def reserve_item(self, item):
        print(f"Inventory: reserved '{item}'")
        return True


class PaymentService:
    def charge(self, amount):
        print(f"Payment: charged ${amount}")
        return True


class ShippingService:
    def schedule_shipment(self, item):
        print(f"Shipping: scheduled shipment for '{item}'")


class NotificationService:
    def send_confirmation(self, item):
        print(f"Notification: confirmation email sent for '{item}'")


class OrderFacade:
    def __init__(self):
        self.inventory = InventoryService()
        self.payment = PaymentService()
        self.shipping = ShippingService()
        self.notification = NotificationService()

    def place_order(self, item, price):
        print(f"--- Placing order for {item} ---")
        if self.inventory.reserve_item(item) and self.payment.charge(price):
            self.shipping.schedule_shipment(item)
            self.notification.send_confirmation(item)
            print("Order complete.")


if __name__ == "__main__":
    order_facade = OrderFacade()
    order_facade.place_order("Wireless Mouse", 24.99)
